File: DEG-main/devkits/p2p-trading-ies-wave1/beckn-signing-kit/nodejs/verifier.js
'use strict';

const crypto = require('crypto');

// SPKI DER prefix for Ed25519 public key (wraps 32-byte raw key).
const SPKI_ED25519_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');

/**
 * Verify that an Authorization header is a valid signature for the given body,
 * using the sender's base64-encoded public key. Throws on failure.
 *
 * @param {Buffer|string} body           - Raw JSON payload
 * @param {string}        authHeader     - Authorization header value
 * @param {string}        publicKeyBase64 - Base64-encoded Ed25519 public key (32 bytes)
 */
function verify(body, authHeader, publicKeyBase64) {
  const nowUnix = Math.floor(Date.now() / 1000);
  verifyAt(body, authHeader, publicKeyBase64, nowUnix);
}

/**
 * Verify at a specific Unix timestamp (useful for testing).
 *
 * @param {Buffer|string} body
 * @param {string}        authHeader
 * @param {string}        publicKeyBase64
 * @param {number}        nowUnix - Unix timestamp in seconds
 */
function verifyAt(body, authHeader, publicKeyBase64, nowUnix) {
  const { created, expires, signature } = parseAuthHeader(authHeader);

  if (created > nowUnix) {
    throw new Error(`verifier: signature not yet valid (created ${created} > now ${nowUnix})`);
  }
  if (nowUnix > expires) {
    throw new Error(`verifier: signature expired (expires ${expires} < now ${nowUnix})`);
  }

  let sigBytes;
  try {
    sigBytes = Buffer.from(signature, 'base64');
  } catch (e) {
    throw new Error(`verifier: invalid signature base64: ${e.message}`);
  }

  const bodyBuf = Buffer.isBuffer(body) ? body : Buffer.from(body);
  const signingString = buildSigningString(bodyBuf, created, expires);

  let pubKeyBytes;
  try {
    pubKeyBytes = Buffer.from(publicKeyBase64, 'base64');
  } catch (e) {
    throw new Error(`verifier: invalid public key base64: ${e.message}`);
  }

  const publicKey = crypto.createPublicKey({
    key: Buffer.concat([SPKI_ED25519_PREFIX, pubKeyBytes]),
    format: 'der',
    type: 'spki',
  });

  const valid = crypto.verify(null, Buffer.from(signingString), publicKey, sigBytes);
  if (!valid) {
    throw new Error('verifier: signature verification failed');
  }
}

/**
 * Extract subscriberID, uniqueKeyID, and algorithm from a keyId field.
 * Format: "subscriber_id|unique_key_id|algorithm"
 *
 * @param {string} keyId
 * @returns {{ subscriberId: string, uniqueKeyId: string, algorithm: string }}
 */
function parseKeyId(keyId) {
  const parts = keyId.split('|');
  if (parts.length !== 3) {
    throw new Error(`invalid keyId format, expected 'subscriber|keyId|algorithm', got "${keyId}"`);
  }
  return {
    subscriberId: parts[0],
    uniqueKeyId: parts[1],
    algorithm: parts[2],
  };
}

/**
 * Build the canonical signing string.
 */
function buildSigningString(body, createdAt, expiresAt) {
  const hash = crypto.createHash('blake2b512');
  hash.update(body);
  const digest = hash.digest('base64');

  return `(created): ${createdAt}\n(expires): ${expiresAt}\ndigest: BLAKE-512=${digest}`;
}

/**
 * Parse the Authorization header to extract created, expires, and signature.
 */
function parseAuthHeader(header) {
  const stripped = header.startsWith('Signature ') ? header.slice('Signature '.length) : header;

  const params = {};
  for (const part of stripped.split(',')) {
    const eqIdx = part.indexOf('=');
    if (eqIdx !== -1) {
      const key = part.slice(0, eqIdx).trim();
      const val = part.slice(eqIdx + 1).replace(/^"|"$/g, '');
      params[key] = val;
    }
  }

  if (!params.created) {
    throw new Error("missing 'created' in auth header");
  }
  if (!params.expires) {
    throw new Error("missing 'expires' in auth header");
  }
  if (!params.signature) {
    throw new Error("missing 'signature' in auth header");
  }

  const created = parseInt(params.created, 10);
  if (isNaN(created)) {
    throw new Error(`invalid 'created' timestamp: ${params.created}`);
  }

  const expires = parseInt(params.expires, 10);
  if (isNaN(expires)) {
    throw new Error(`invalid 'expires' timestamp: ${params.expires}`);
  }

  return { created, expires, signature: params.signature };
}

module.exports = { verify, verifyAt, parseKeyId };

File: DEG-main/devkits/p2p-trading-ies-wave1/beckn-signing-kit/golang/verifier.go
package signer

import (
	"crypto/ed25519"
	"encoding/base64"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Verify checks that an Authorization header is a valid signature for the given body,
// using the sender's base64-encoded public key. Returns nil if valid.
func Verify(body []byte, authHeader string, publicKeyBase64 string) error {
	return VerifyAt(body, authHeader, publicKeyBase64, time.Now())
}

// VerifyAt checks validity at a specific time (useful for testing).
func VerifyAt(body []byte, authHeader string, publicKeyBase64 string, now time.Time) error {
	created, expires, sig, err := parseAuthHeader(authHeader)
	if err != nil {
		return fmt.Errorf("verifier: %w", err)
	}

	currentTime := now.Unix()
	if created > currentTime {
		return fmt.Errorf("verifier: signature not yet valid (created %d > now %d)", created, currentTime)
	}
	if currentTime > expires {
		return fmt.Errorf("verifier: signature expired (expires %d < now %d)", expires, currentTime)
	}

	sigBytes, err := base64.StdEncoding.DecodeString(sig)
	if err != nil {
		return fmt.Errorf("verifier: invalid signature base64: %w", err)
	}

	signingString, err := buildSigningString(body, created, expires)
	if err != nil {
		return fmt.Errorf("verifier: %w", err)
	}

	pubKeyBytes, err := base64.StdEncoding.DecodeString(publicKeyBase64)
	if err != nil {
		return fmt.Errorf("verifier: invalid public key base64: %w", err)
	}

	if !ed25519.Verify(ed25519.PublicKey(pubKeyBytes), []byte(signingString), sigBytes) {
		return fmt.Errorf("verifier: signature verification failed")
	}

	return nil
}

// ParseKeyID extracts subscriberID, uniqueKeyID, and algorithm from a keyId field.
// Format: "subscriber_id|unique_key_id|algorithm"
func ParseKeyID(keyID string) (subscriberID, uniqueKeyID, algorithm string, err error) {
	parts := strings.Split(keyID, "|")
	if len(parts) != 3 {
		return "", "", "", fmt.Errorf("invalid keyId format, expected 'subscriber|keyId|algorithm', got %q", keyID)
	}
	return parts[0], parts[1], parts[2], nil
}

func parseAuthHeader(header string) (created, expires int64, signature string, err error) {
	header = strings.TrimPrefix(header, "Signature ")

	params := make(map[string]string)
	for _, part := range strings.Split(header, ",") {
		kv := strings.SplitN(strings.TrimSpace(part), "=", 2)
		if len(kv) == 2 {
			params[strings.TrimSpace(kv[0])] = strings.Trim(kv[1], `"`)
		}
	}

	createdStr, ok := params["created"]
	if !ok {
		return 0, 0, "", fmt.Errorf("missing 'created' in auth header")
	}
	created, err = strconv.ParseInt(createdStr, 10, 64)
	if err != nil {
		return 0, 0, "", fmt.Errorf("invalid 'created' timestamp: %w", err)
	}

	expiresStr, ok := params["expires"]
	if !ok {
		return 0, 0, "", fmt.Errorf("missing 'expires' in auth header")
	}
	expires, err = strconv.ParseInt(expiresStr, 10, 64)
	if err != nil {
		return 0, 0, "", fmt.Errorf("invalid 'expires' timestamp: %w", err)
	}

	signature, ok = params["signature"]
	if !ok {
		return 0, 0, "", fmt.Errorf("missing 'signature' in auth header")
	}

	return created, expires, signature, nil
}
